### PR TITLE
Allowing passing a null pointer to getrandom() when length is 0

### PR DESCRIPTION
**Pull request by Aaron1011**
The Linux kernel will handle a null pointer passed to 'getrandom'
without error, as long as the length is also 0. The `getrandom` crate
relies on this behavior: https://github.com/rust-random/getrandom/blob/ab44edf3c7af721a00e22648b6c811ccb559ba81/src/linux_android.rs#L26

Since it works fine on the actual kernel (and should continue to, due to
the kernel's backwards-compatibility guarantees), Miri should support it
as well.

### CODE REVIEW

**Comment by RalfJung** [comment]
We already handle NULL pointers just fine when the length is 0. The `not_undef` just checks that the pointer is *initialized*, and we should definitely still do that.

I think you ran into the same getrandom bug that I also just ran into. See https://github.com/rust-random/getrandom/pull/73.

**Comment by RalfJung** [comment]
FWIW, even with this PR, getrandom still fails here. (You can do `cargo update && cargo miri test` in `test-cargo-miri/` to reproduce.)

**Comment by Aaron1011** [comment]
@RalfJung: Wow - I opened a duplicate PR within a few minutes of yours: https://github.com/rust-random/getrandom/pull/74

However, applying either of our PRs is not sufficient to make that call to `getrandom` work. I think the blame actually lies with `gen_random`:

https://github.com/rust-lang/miri/blob/ed30152b52f650be4e8e2eee8616f7fa41f89569/src/helpers.rs#L86

We unconditionally try to use that pointer. I think we should add a flag to skip doing anything when the length is 0. We can enable that behavior for `getrandom`, and keep the current behavior for all other usages of `gen_random`.

**Comment by RalfJung** [comment]
> I opened a duplicate PR within a few minutes of yours: rust-random/getrandom#74

Talk about race conditions. ;)

> However, applying either of our PRs is not sufficient to make that call to getrandom work. I think the blame actually lies with gen_random:

Ah, good catch. `check_ptr_access` complains on NULL. So *dangling* pointers (and unaligned pointers) work fine here with size 0, but NULL pointers do not.

I find it likely that all users of `gen_random` are fine with NULL pointers, so I think `gen_random` itself should be changed to bail early if the size is 0.

**Comment by Aaron1011** [comment]
@RalfJung: I've made the change you requested

File: src/helpers.rs [hunk 1]
```diff
@@ -81,6 +81,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
+        // to their platform's random-generation function (e.g. getrandom())
+        // on Linux. For compatibility with these programs, we don't perform
+        // any additional checks - it's okay if the pointer is invalid,
+        // since we wouldn't actually be writing to it.
+        if len == 0 {
+            return Ok(())
+        }
         let this = self.eval_context_mut();
 
         let ptr = match this.memory().check_ptr_access(ptr, Size::from_bytes(len as u64), Align::from_bytes(1).unwrap())? {
```
Comment (RalfJung) [inline src/helpers.rs]: This match can now be replaced by an `.expect("we already checked for size 0")`

File: src/helpers.rs [hunk 2]
```diff
@@ -81,6 +81,14 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         ptr: Scalar<Tag>,
         len: usize,
     ) -> InterpResult<'tcx>  {
+        // Some programs pass in a null pointer and a length of 0
+        // to their platform's random-generation function (e.g. getrandom())
+        // on Linux. For compatibility with these programs, we don't perform
+        // any additional checks - it's okay if the pointer is invalid,
+        // since we wouldn't actually be writing to it.
+        if len == 0 {
+            return Ok(())
```
Comment (RalfJung) [inline src/helpers.rs]:
```suggestion
            return Ok(());
```
I think that's the style we are usually using.

File: src/helpers.rs [hunk 3]
```diff
@@ -91,10 +91,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         let this = self.eval_context_mut();
 
-        let ptr = match this.memory().check_ptr_access(ptr, Size::from_bytes(len as u64), Align::from_bytes(1).unwrap())? {
-            Some(ptr) => ptr,
-            None => return Ok(()), // zero-sized access
-        };
+        let ptr = this.memory().check_ptr_access(ptr, Size::from_bytes(len as u64), Align::from_bytes(1).unwrap())?.expect("we already checked for size 0");
```
Comment (RalfJung) [inline src/helpers.rs]: Please break the long line (I know it was wide before, and we don't have a strict 100 column limit in Miri, but `expect` really is very wide)

**Comment by Aaron1011** [comment]
@RalfJung Updated

**Comment by RalfJung** [comment]
Thanks!

@bors r+

**Pull request merged**